Makefile: new target kind-debug to debug cilium operator & agent in kind cluster

mhofstetter opened the pull request:
A new make target `kind-debug` has been added. In addition to `make kind-debug-agent` it adds the possibility to debug the cilium operator beside the agents.

It follows the same ideas which has been introduced for the cilium agent in https://github.com/cilium/cilium/pull/21108

To allow easy remote-debugging of the operator, the following points has been addressed:
* operator gets pinned to kind-worker with nodeSelector
* operator debug port gets exposed on every kind node (for the sake of completeness) (controlplane: 2350x, worker: 2351x)(possible due to podAntiAffinity -> only one operator pod per node anyway)
* operator HA mode isn't addressed (operator doesn't get deployed this way anyway)
* Use of the delve flag `--continue` to remove the need to attach a debugger to all instances. (deactivate by setting environment variable `DEBUG_HOLD=true make kind-debug`)

mhofstetter committed:
.nvim: Add cilium operator debug target

This commit adds the debug target for debugging the cilium operator
within a k8s kind cluster.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
.vscode: Add cilium operator debug target

This commit adds the debug target for debugging the cilium operator
within a k8s kind cluster.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
images: configurable debug port in debug-wrapper.sh

This commit adds the possibility to configure the debug port as
environment variable DEBUG_PORT. It still defaults to 2345.

This adds the possibility to use the debug-wrapper for the cilium
operator which needs to expose the debugger on a different port than the
agent.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
images: optionally continue delve debugger at startup via flag --continue

By passing the environment variable DEBUG_CONTINUE=true to the make
target, the devle flag --continue is set. This way, the cilium agent
& operator don't wait for an attached debugger.

Without this change, it might gets a tedious action to attach the
debugger to all pods.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
Makefile: add make target kind-debug

This commit adds a new make target kind-debug which builds & deploys the
cilium agent & operator in a local kind cluster for debugging with
delve.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
Makefile: silence output of creating .dockerignore files

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
contrib: Expose cilium operator debug port in kind setup

With this commit, the operator debug port gets exposed on every kind
node. In addition, the operator pod gets pinned to node kind-worker (via
a nodeSelector on the hostname) which eases debuggability. This is
possible due to the default podAntiAffinity, which ensures, that only
one operator pod is running on one k8s node.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
images: docker target debug for cilium operator

This commit introduces a new docker target debug which provides the
possibility to build a cilium operator image which includes the
debug-wrapper.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

joestringer commented on images/operator/Dockerfile:
```diff
@@ -69,3 +71,35 @@ COPY --from=builder /out/${TARGETOS}/${TARGETARCH}/LICENSE.all /LICENSE.all
+# image, install delve debugger and wrap the cilium-operator binary calls
+# with a script that automatically provisions the debugger on a
+# dedicated port.
+FROM --platform=${BUILDPLATFORM} ${CILIUM_RUNTIME_IMAGE} as debug
```
Why `${CILIUM_RUNTIME_IMAGE}` rather than `release`? The makefiles should already disable stripping etc, so we should be able to avoid duplicating the COPYs etc.

mhofstetter replied:
`release` is `scratch` in case of the operator - therefore not containing `bash` - which is necessary for the `debug-wrapper`. 

that's the reason why i decided to rebuild the debug image on top of the runtime image - which brings the duplication.

another option would be change `release` to `runtime-image` (don't prefer this one) or copying `bash` into the debug image (if this is possible).

mhofstetter replied:
replacing the `debug-wrapper.sh` with a statically linked go binary would also eliminate the requirement for having bash available in the debug image. i'm kind of in favor with this approach :thinking: 

having the binary built in the cilium-builder image (alongside installing delve) and just copying it afterwards into the respective debug images.

@joestringer what do you think?

joestringer replied:
I'd be open to replacing debug-wrapper.sh if you have a good idea how to do so. My main motivations with the current approach were just to change as few things as possible from the existing release images&helm charts.

It didn't occur to me to statically link the binary against delve to introduce the debugger, the wrapper script was just the minimal way to get that done.

mhofstetter replied:
* replaced the bash script with a small and simple go app which gets built in the builder image
* had to install delve with `CGO=0`

with this changes in place, it possible to build the debug image of the operator on top of its release image and only add the necessary stuff (copy delve, copy debug wrapper, re-adding operator binary itself (because `mv` isn't available in `scratch` too))

this way the diff between `release` & `debug` and `agent` & `operator` are minimal even though the operator is built on top of `scratch`

joestringer commented on images/operator/Dockerfile:
```diff
@@ -69,3 +71,35 @@ COPY --from=builder /out/${TARGETOS}/${TARGETARCH}/LICENSE.all /LICENSE.all
 CMD ["/usr/bin/cilium-${OPERATOR_VARIANT}"]
+
+# Intermediate stage because expansion is not working in COPY --from
+FROM --platform=${BUILDPLATFORM} ${CILIUM_BUILDER_IMAGE} as debug-builder
```
:thinking: Why isn't expansion working in `COPY --from`? What's different in this image vs. `images/cilium/Dockerfile`?

mhofstetter replied:
`COPY --from=${CILIUM_BUILDER_IMAGE} /go/bin/dlv /usr/bin/dlv`

results in 

```
Dockerfile:98
--------------------
  96 |     CMD ["/usr/bin/cilium-${OPERATOR_VARIANT}"]
  97 |     
  98 | >>> COPY --from=${CILIUM_BUILDER_IMAGE} /go/bin/dlv /usr/bin/dlv
  99 |     RUN mv /usr/bin/cilium-${OPERATOR_VARIANT} /usr/bin/cilium-${OPERATOR_VARIANT}-bin
 100 |     ENV DEBUG_PORT=2346
--------------------
ERROR: failed to solve: failed to parse stage name "${CILIUM_BUILDER_IMAGE}": invalid reference format: repository name must be lowercase
```

even though the following would work (but would break the cilium runtime/builder image update logic.

`COPY --from=quay.io/cilium/cilium-builder:f55691636922518ca40802f59d57f36ee4741c0d@sha256:4c21a3e21e9e07f0301191b3ae6921c8208a3910ce1b73be6ad445e544590e3a /go/bin/dlv /usr/bin/dlv`

the builder stage of `cilium/Dockerfile` already builds on top of the cilium builder image - there doesn't have the exact same need.

but the dockerfile of cilium builder itself also has to redefine all images as stages to be able to copy files in other stages. (https://github.com/cilium/cilium/blob/master/images/builder/Dockerfile#L4-L16)

joestringer replied:
In the cilium Dockerfile, we use `COPY --from=builder ...` rather than referencing a variable. Could we just do that?

joestringer replied:
(overall I'm not too worried, I'm just thinking that it'd be good to minimize the differences between the release build vs. debug builds, and reduce duplicated statements. If we ever change things, it's easy to accidentally make the release vs. debug build diverge)

mhofstetter replied:
> In the cilium Dockerfile, we use COPY --from=builder ... rather than referencing a variable. Could we just do that?

well the `builder` is just a build stage itself - which uses a variable when being defined too. so actually, no difference between the operator and agent.

> good to minimize the differences between the release build vs. debug builds, and reduce duplicated statements

strongly agree on that one. going to rename the stage from `debug-builder` to `debug-tools` where we then copy dlv & debug-wrapper from - only in case of the debugger image of course. should be fine then.

joestringer commented on images/scripts/debug-wrapper.sh:
```diff
@@ -4,9 +4,13 @@ set -o errexit
-    --listen=:2345 \
+    --listen=":${DEBUG_PORT}" \
     --headless=true \
+    --continue="${DEBUG_CONTINUE}" \
```
I like the continue step. I wonder whether it's more useful to developers to turn this on by default or off by default? For the `kind-image` target it's a bit annoying to have to go through and release each debugger any time you restart the agent or operator pods, so maybe the default should be to continue and we should instead do a `DEBUG_HOLD` var?

(maybe worth raising this in #development on Slack to see if other devs have alternate opinions on this front)

mhofstetter replied:
personally, i would prefer the same. 

but i wanted to introduce this change without changing the existing logic. which is also the reason why i decided to keep `make kind-debug-agent` for the time being.

an option would be to set `DEBUG_CONTINUE` only in case of the newly introduced make target `kind-debug` and keep it as is in `kind-debug-agent`

joestringer replied:
I suggest that we pick a proposal (I would suggest continue by default), and use the Slack and/or community meeting to share that proposal so that anyone who cares can voice their thoughts on this. It would also help others to know about (a) this debug feature overall and (b) that the behaviour will change. I had introduced the change originally and I'm on board with changing it. In this case since we're talking about changing behaviour for developers, I don't think we need to be shy about changing the behaviour.

I would prefer if we don't conditionally change the behaviour depending on the target. That will make it more confusing for anyone new who attempts to use one of the targets and switches between the targets for some reason. Let's make it consistent and pick a default that works well for 90% of cases, then allow developers to override it if they have a special case they want for their own workflow. IMO setting continue by default is the right choice here.

mhofstetter replied:
:+1: 

going to 
* switch the default to `continue=true` (our current recommendation)
* rename the env variable for changing the default to `DEBUG_HOLD`
* keep the targets `kind-debug` &  `kind-debug-agent` the same regarding this default behaviour

will bring this up in slack/community meeting next week

joestringer reviewed (commented):
Nice work, thanks! I have a couple of questions about the details in the operator images where ideally we could reduce the amount of duplicate logic. Then there's one more discussion point I'm not sure about the resolution yet but where it may help to have additional input from the developer community.

mhofstetter committed:
images: install delve without CGO

Delve should be go installed with CGO=0 in the cilium builder image.
This enables to use the binary in images based upon scratch (e.g. cilium
operator)

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
images: build debug-wrapper binary in cilium builder image

the debug wrapper go binary should be built and put into the builder
image alongside installing the delve binary.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
images: remove debug-wrapper based on bash

the debug-wrapper based on bash has been replaced with a statically
linked go binary which is also usable in images not having bash on its
PATH.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
images: using go binary of debug-wrapper in agent & operator images

By replacing the bash-based debug-wrapper with a statically linked go
binary - the debug image of the operator can base upon the release stage
which doesn't include bash (scratch).

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
images: continue delve debugger by default at startup

By default, the debugger continues at startup. If this is undesired, the
environment variable DEBUG_HOLD=true can be set to build the image with
the delve flag --continue=false.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

joestringer commented on images/builder/debug-wrapper.go:
```diff
@@ -0,0 +1,54 @@
+		os.Exit(1)
+	}
+
+	inputArts := os.Args[1:]
```
nit, seems like a typo (though consistent elsewhere :sweat_smile: )
```suggestion
	inputArgs := os.Args[1:]
```

mhofstetter replied:
thanks, fixed

joestringer commented on images/builder/debug-wrapper.go:
```diff
@@ -0,0 +1,54 @@
+	cmd.Stderr = os.Stderr
+
+	if err = cmd.Run(); err != nil {
+		log.Printf("failed to execute dlv: %v", err)
```
Do we need to filter the error types here, such as for `*exec.ExitError`? The error that is printed to the stdout here may be a bit confusing if everything is successful except that cilium-agent returns with exit code 1 for instance.

mhofstetter replied:
good point. refactored the part that `ExitErrors` are handled separately by suppressing a log message and returning the exit code.

all other errors are handled by logging it and returning with exit code 1

joestringer commented on images/operator/Dockerfile:
```diff
@@ -2,6 +2,8 @@
 
 ARG BASE_IMAGE=scratch
+ARG CILIUM_BUILDER_IMAGE=quay.io/cilium/cilium-builder:f55691636922518ca40802f59d57f36ee4741c0d@sha256:4c21a3e21e9e07f0301191b3ae6921c8208a3910ce1b73be6ad445e544590e3a
+ARG CILIUM_RUNTIME_IMAGE=quay.io/cilium/cilium-runtime:b11a6777dec1d4096dacd3fd1b94090c62133192@sha256:b2a3bf605efb9b89d6684fabb76d841bc9d7a85555ba5bdf7945409334df1779
```
`CILIUM_RUNTIME_IMAGE` appears to be unused?
```suggestion
```

mhofstetter replied:
thanks, removed

joestringer commented on images/operator/Dockerfile:
```diff
@@ -2,6 +2,8 @@
 # SPDX-License-Identifier: Apache-2.0
 
 ARG BASE_IMAGE=scratch
+ARG CILIUM_BUILDER_IMAGE=quay.io/cilium/cilium-builder:f55691636922518ca40802f59d57f36ee4741c0d@sha256:4c21a3e21e9e07f0301191b3ae6921c8208a3910ce1b73be6ad445e544590e3a
```
This image sha differs from the one in `images/cilium/Dockerfile`. Hmm. Usually if we change the cilium builder or runtime images, we need to push the PR to the cilium/cilium repo directly and then there's a bot that will autobuild the runtime/builder images and update these SHAs. I suspect that if we merged this PR as-is, something may break but I'm not entirely sure.

@aanm what do you think the best workflow will be here? Strictly speaking the new code doesn't need to be in the builder image, but it won't change often so it's kinda nice to build it once generically so that it can be reused in both the operator and cilium images.

aanm replied:
@joestringer the `make -C images/ update-builder-image` will pick these changes automatically. No need to change anything in this PR.

joestringer replied:
OK, so you're saying we can just ignore the failing "[Base Image Lint](https://github.com/cilium/cilium/actions/runs/4251959190/jobs/7394959081)" job below, merge this once it's passed review, then immediately open a PR to update the builder image using the above?

joestringer replied:
Or are you saying that once [this build](https://github.com/cilium/cilium/actions/runs/4251959003) completes, Marco can just run the make command above to pull in the right image version?

mhofstetter replied:
rebased to master to pull in the latest runtime/builder images which seems to be changed on master. this should fix the base image lint issue and hopefully let me trigger the builder image build with my changes :pray:

mhofstetter replied:
cilium-builder image built successfully

pushed a commit which changes the usages to the corresponding image

joestringer reviewed (changes requested):
Just some minor nits but the latest version looks good overall. I'm just not 100% sure about how to ensure these changes end up in the canonical builder images and how we can ensure all dependent images are synced up properly.

(I also didn't manually test this myself)

aanm commented on images/operator/Dockerfile:
```diff
@@ -2,6 +2,8 @@
 # SPDX-License-Identifier: Apache-2.0
 
 ARG BASE_IMAGE=scratch
+ARG CILIUM_BUILDER_IMAGE=quay.io/cilium/cilium-builder:f55691636922518ca40802f59d57f36ee4741c0d@sha256:4c21a3e21e9e07f0301191b3ae6921c8208a3910ce1b73be6ad445e544590e3a
```
Can we move this ARG to line 74? Just to make it more obvious that this image is not used in production.

mhofstetter replied:
unfortunately this seems not to be possible :thinking: 

moving it results in the following error. the variable needs to be defined at the beginning (globally) to be available in the `FROM` statements when defining a new layer.

```
Dockerfile:74
--------------------
  72 |     
  73 |     ARG CILIUM_BUILDER_IMAGE=quay.io/cilium/cilium-builder:6615810f9d5595ea9639e91255bdbe83acb4b5fb@sha256:595c6a8c06db1dd2a71a4a2891b270af674705b7d1f7a2cfec5bfac64bd2af74
  74 | >>> FROM --platform=${BUILDPLATFORM} ${CILIUM_BUILDER_IMAGE} as debug-tools
  75 |     
  76 |     #
--------------------
ERROR: failed to solve: base name (${CILIUM_BUILDER_IMAGE}) should not be blank

```

aanm replied:
It's unfortunate.

mhofstetter committed:
images: update cilium-{runtime,builder}

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

joestringer reviewed (approved):
LGTM thanks!

tklauser commented on images/operator/Dockerfile:
```diff
@@ -69,3 +70,26 @@ COPY --from=builder /out/${TARGETOS}/${TARGETARCH}/LICENSE.all /LICENSE.all
+# Typical image bulids will stop above at the 'release' target, but
+# developers follow this Dockerfile to the end. Starting from a release
+# image, install delve debugger and wrap the cilium-operator binary calls
+# with a script that automatically provisions the debugger on a
```
Super nit:

```suggestion
# with the statically linked debug-wrapper binary that automatically provisions the debugger on a
```

Only worth addressing in case there are other change requests.

tklauser reviewed (approved):
Thanks!
